### DISCUSSION TITLE
io/netty/resolver/dns/DnsAddressResolverGroup/DnsAddressResolverGroup is not assignable to io/netty/resolver/AddressResolverGroup

### DISCUSSION THREAD

**Discussion by klnlakshmiphilips** [redis/lettuce#2916] [Q&A]
## Bug Report

When i try to intialize the redis cluster client with the below code, i get exception as described below.

```java
  RedisURI redisUriCluster = RedisURI.builder().withHost(host).withPort(port).build();
    redisClusterClient = RedisClusterClient.create(redisUriCluster);
```

```
Exception Details:
  Location:
    io/lettuce/core/resource/AddressResolverGroupProvider$DefaultDnsAddressResolverGroupWrapper.<clinit>()V @51: putstatic
  Reason:
    Type 'io/netty/resolver/dns/DnsAddressResolverGroup' (current frame, stack[0]) is not assignable to 'io/netty/resolver/AddressResolverGroup'
```

This exception is triggered from the class io.lettuce.core.resource.AddressResolverGroupProvider in the static block.

The static block have the following code

```java
static {
        boolean dnsResolverAvailable = LettuceClassUtils.isPresent("io.netty.resolver.dns.DnsAddressResolverGroup");
        Object group;
        if (dnsResolverAvailable) {
            logger.debug("Starting with netty's non-blocking DNS resolver library");
            group = AddressResolverGroupProvider.DefaultDnsAddressResolverGroupWrapper.INSTANCE;
        } else {
            logger.debug("Starting without optional netty's non-blocking DNS resolver library");
            group = DefaultAddressResolverGroup.INSTANCE;
        }

        ADDRESS_RESOLVER_GROUP = (AddressResolverGroup)group;
    }
```

As per this code when the class "io.netty.resolver.dns.DnsAddressResolverGroup" is part of the classapath and java is able to load the class, then the boolean value will be true.

As i am running my flink job in EMR the hadoop have a jar netty-resolver-dns-4.1.77.Final.jar in the class path. And this Jar have the class "io.netty.resolver.dns.DnsAddressResolverGroup". After the condition is true, the following code gets executed and throws the above exception

```java
private static class DefaultDnsAddressResolverGroupWrapper {
        static AddressResolverGroup<?> INSTANCE = new DnsAddressResolverGroup((new DnsNameResolverBuilder()).channelType(Transports.datagramChannelClass()).socketChannelType(Transports.socketChannelClass().asSubclass(SocketChannel.class)).cnameCache(new DefaultDnsCnameCache()).resolveCache(new DefaultDnsCache()));

        private DefaultDnsAddressResolverGroupWrapper() {
        }
    }
```

#### Current Behavior

<details>
<summary>Stack trace</summary>

Exception Details:
  Location:
    io/lettuce/core/resource/AddressResolverGroupProvider$DefaultDnsAddressResolverGroupWrapper.<clinit>()V @51: putstatic
  Reason:
    Type 'io/netty/resolver/dns/DnsAddressResolverGroup' (current frame, stack[0]) is not assignable to 'io/netty/resolver/AddressResolverGroup'
  Current Frame:
    bci: @51
    flags: { }
    locals: { }
    stack: { 'io/netty/resolver/dns/DnsAddressResolverGroup' }

</details>

#### Input Code

<details>
<summary>Input Code</summary>

  RedisURI redisUriCluster = RedisURI.builder().withHost(host).withPort(port).build();
    redisClusterClient = RedisClusterClient.create(redisUriCluster);

</details>

#### Environment

- Lettuce version(s): 6.3.1.RELEASE
- Redis version: AWS ElasticCache for Redis 7.1.0

**Comment by tishun** [comment]
Hey @klnlakshmiphilips ,

could you verify that the `io.netty:netty-resolver-dns-4.1.77.Final.jar` and the `io.netty:netty-resolver-4.1.77.Final.jar` are both present in the classpath and both share the same version ?
